File: packages/typescript/src/Volume.ts
import { ToolboxApi, VolumeDto, VolumesApi } from '@daytonaio/api-client'
import { SandboxInstance } from './Sandbox'

/**
 * Represents a Daytona Volume which is a shared storage volume for Sandboxes.
 *
 * @property {string} id - Unique identifier for the Volume
 * @property {string} name - Name of the Volume
 * @property {string} organizationId - Organization ID that owns the Volume
 * @property {string} state - Current state of the Volume
 * @property {string} createdAt - Date and time when the Volume was created
 * @property {string} updatedAt - Date and time when the Volume was last updated
 * @property {string} lastUsedAt - Date and time when the Volume was last used
 */
export type Volume = VolumeDto & { __brand: 'Volume' }

/**
 * Service for managing Daytona Volumes.
 *
 * This service provides methods to list, get, create, and delete Volumes.
 *
 * @class
 */
export class VolumeService {
  constructor(private volumesApi: VolumesApi) {}

  /**
   * Lists all available Volumes.
   *
   * @returns {Promise<Volume[]>} List of all Volumes accessible to the user
   *
   * @example
   * const daytona = new Daytona();
   * const volumes = await daytona.volume.list();
   * console.log(`Found ${volumes.length} volumes`);
   * volumes.forEach(vol => console.log(`${vol.name} (${vol.id})`));
   */
  async list(): Promise<Volume[]> {
    const response = await this.volumesApi.listVolumes()
    return response.data as Volume[]
  }

  /**
   * Gets a Volume by its ID.
   *
   * @param {string} id - ID of the Volume to retrieve
   * @returns {Promise<Volume>} The requested Volume
   * @throws {Error} If the Volume does not exist or cannot be accessed
   *
   * @example
   * const daytona = new Daytona();
   * const volume = await daytona.volume.get("volume-uuid");
   * console.log(`Volume ${volume.name} is in state ${volume.state}`);
   */
  // TODO: volume - should be name
  async get(id: string): Promise<Volume> {
    const response = await this.volumesApi.getVolume(id)
    return response.data as Volume
  }

  /**
   * Creates a new Volume with the specified name.
   *
   * @param {string} name - Name for the new Volume
   * @returns {Promise<Volume>} The newly created Volume
   * @throws {Error} If the Volume cannot be created
   *
   * @example
   * const daytona = new Daytona();
   * const volume = await daytona.volume.create("my-data-volume");
   * console.log(`Created volume ${volume.name} with ID ${volume.id}`);
   */
  async create(name: string): Promise<Volume> {
    const response = await this.volumesApi.createVolume({ name })
    return response.data as Volume
  }

  /**
   * Deletes a Volume by its ID.
   *
   * @param {string} id - ID of the Volume to delete
   * @returns {Promise<void>}
   * @throws {Error} If the Volume does not exist or cannot be deleted
   *
   * @example
   * const daytona = new Daytona();
   * await daytona.volume.delete("volume-uuid");
   * console.log("Volume deleted successfully");
   */
  // TODO: volume - should be name
  async delete(id: string): Promise<void> {
    await this.volumesApi.deleteVolume(id)
  }
}

/**
 * Service for managing Volume operations within a Sandbox.
 *
 * This service provides methods to mount and unmount Volumes to specific paths
 * within a Sandbox's filesystem.
 *
 * @class
 */
export class SandboxVolumeService {
  constructor(private instance: SandboxInstance, private toolboxApi: ToolboxApi) {}

  /**
   * Mounts a Volume to a specific path in the Sandbox.
   *
   * This makes the Volume's contents available at the specified path
   * in the Sandbox's filesystem.
   *
   * @param {string} volumeId - ID of the Volume to mount
   * @param {string} path - Path in the Sandbox where the Volume should be mounted
   * @returns {Promise<void>}
   * @throws {Error} If the Volume cannot be mounted
   *
   * @example
   * const daytona = new Daytona();
   * const sandbox = await daytona.create();
   * await sandbox.volume.mount("volume-uuid", "/mnt/data");
   */
  // TODO: volume - should be volume name
  async mount(volumeId: string, path: string): Promise<void> {
    await this.toolboxApi.mountVolume(this.instance.id, { volumeId, path })
  }

  /**
   * Unmounts a Volume from the Sandbox.
   *
   * This removes the Volume from the Sandbox's filesystem.
   * Any data written to the Volume while mounted will be preserved.
   *
   * @param {string} volumeId - ID of the Volume to unmount
   * @returns {Promise<void>}
   * @throws {Error} If the Volume cannot be unmounted
   *
   * @example
   * const daytona = new Daytona();
   * const sandbox = await daytona.create();
   *
   * // Mount volume
   * await sandbox.volume.mount("volume-uuid", "/mnt/data");
   * // Use volume
   * await sandbox.process.exec("echo 'hello' > /mnt/data/test.txt");
   * // Unmount when done
   * await sandbox.volume.unmount("volume-uuid");
   * console.log(`Volume ${volume.name} unmounted`);
   */
  // TODO: volume - should be volume name
  async unmount(volumeId: string): Promise<void> {
    await this.toolboxApi.unmountVolume(this.instance.id, { volumeId, path: '/' })
  }
}

File: packages/python/src/daytona_sdk/volume.py
from typing import List

from daytona_api_client import (
    CreateVolume,
    ToolboxApi,
    VolumeDto,
    VolumeMountDto,
    VolumesApi,
)

from .protocols import SandboxInstance


class Volume(VolumeDto):
    """Represents a Daytona Volume which is a shared storage volume for Sandboxes.

    Attributes:
        id (StrictStr): Unique identifier for the Volume.
        name (StrictStr): Name of the Volume.
        organization_id (StrictStr): Organization ID of the Volume.
        state (StrictStr): State of the Volume.
        created_at (StrictStr): Date and time when the Volume was created.
        updated_at (StrictStr): Date and time when the Volume was last updated.
        last_used_at (StrictStr): Date and time when the Volume was last used.
    """

    @classmethod
    def from_dto(cls, dto: VolumeDto) -> "Volume":
        return cls(**dto.__dict__)


class VolumeService:
    """Service for managing Daytona Volumes. Can be used to list, get, create and delete Volumes."""

    def __init__(self, volumes_api: VolumesApi):
        self.__volumes_api = volumes_api

    def list(self) -> List[Volume]:
        """List all Volumes.

        Returns:
            List[Volume]: List of all Volumes.

        Example:
            ```python
            daytona = Daytona()
            volumes = daytona.volume.list()
            print(volumes)
            ```
        """
        return [Volume.from_dto(volume) for volume in self.__volumes_api.list_volumes()]

    def get(
        self, id: str
    ) -> Volume:  # TODO: volume - should be name # pylint: disable=fixme
        """Get a Volume by ID.

        Args:
            id (str): ID of the Volume to get.

        Returns:
            Volume: The Volume object.

        Example:
            ```python
            daytona = Daytona()
            volume = daytona.volume.get("test-volume-uuid")
            print(volume)
            ```
        """
        return Volume.from_dto(self.__volumes_api.get_volume(id))

    def create(self, name: str) -> Volume:
        """Create a new Volume.

        Args:
            name (str): Name of the Volume to create.

        Returns:
            Volume: The Volume object.

        Example:
            ```python
            daytona = Daytona()
            volume = daytona.volume.create("test-volume")
            print(volume)
            ```
        """
        return Volume.from_dto(
            self.__volumes_api.create_volume(CreateVolume(name=name))
        )

    def delete(
        self, id: str
    ) -> None:  # TODO: volume - should be name # pylint: disable=fixme
        """Delete a Volume by ID.

        Args:
            id (str): ID of the Volume to delete.

        Example:
            ```python
            daytona = Daytona()
            daytona.volume.delete("test-volume-uuid")
            ```
        """
        self.__volumes_api.delete_volume(id)


class SandboxVolumeService:
    """Service for managing Daytona Volumes in a Sandbox. Can be used to mount and unmount Volumes from a Sandbox."""

    def __init__(self, instance: SandboxInstance, toolbox_api: ToolboxApi):
        self.__instance = instance
        self.__toolbox_api = toolbox_api

    def mount(
        self, volume_id: str, path: str
    ) -> None:  # TODO: volume - should be volume name # pylint: disable=fixme
        """Mount a Volume to a Sandbox.

        Args:
            volume_id (str): ID of the Volume to mount.
            path (str): Path to mount the Volume to.

        Example:
            ```python
            daytona = Daytona()
            sandbox = daytona.create()
            sandbox.volume.mount("test-volume-uuid", "/mnt/test")
            ```
        """
        self.__toolbox_api.mount_volume(
            self.__instance.id, VolumeMountDto(volume_id=volume_id, path=path)
        )

    def unmount(
        self, volume_id: str
    ) -> None:  # TODO: volume - should be volume name # pylint: disable=fixme
        """Unmount a Volume from a Sandbox.

        Args:
            volume_id (str): ID of the Volume to unmount.

        Example:
            ```python
            daytona = Daytona()
            sandbox = daytona.create()
            sandbox.volume.mount("test-volume-uuid", "/mnt/test")
            # do some work
            sandbox.volume.unmount("test-volume-uuid")
            ```
        """
        self.__toolbox_api.unmount_volume(
            self.__instance.id, VolumeMountDto(volume_id=volume_id, path="/")
        )
